dipy/denoise/enhancement_kernel.pyx

```diff
@@ -0,0 +1,420 @@
+                                arg2[0], arg2[1])
+        kernelval = self.kernel(c)
+
+        return kernelval
+
+    @cython.wraparound(False)
+    @cython.boundscheck(False)
+    @cython.nonecheck(False)
+    @cython.cdivision(True)
+    cdef double [:] coordinate_map(self, double x, double y,
+                                   double z, double beta,
+                                   double gamma) nogil:
+        """ Compute a coordinate map for the kernel
+
+        Parameters
+        ----------
+        x : double
+            X position
+        y : double
+            Y position
+        z : double
+            Z position
+        beta : double
+            First Euler angle
+        gamma : double
+            Second Euler angle
+
+        Returns
+        -------
+        c : array of double
+            array of coordinates for kernel
+        """
+
+        cdef:
+            double [:] c
+            double q
+            double cg
+            double cotq2
+
+        with gil:
+
+            c = np.zeros(6)
+
+        if beta == 0:
+            c[0] = x
+            c[1] = y
+            c[2] = z
+            c[3] = c[4] = c[5] = 0
+
+        else:
+            q = fabs(beta)
+            cg = cos(gamma)
+            sg = sin(gamma)
+            cotq2 = 1.0 / tan(q/2)
+
+            c[0] = -0.5*z*beta*cg + \
+                    x*(1 - (beta*beta*cg*cg * (1 - 0.5*q*cotq2)) / (q*q)) - \
+                    (y*beta*beta*cg*sg * (1 - 0.5*q*cotq2)) / (q*q)
+            c[1] = -0.5*z*beta*sg - \
+                    (x*beta*beta*cg*sg * (1 - 0.5*q*cotq2)) / (q*q) + \
+                    y * (1 - (beta*beta*sg*sg * (1 - 0.5*q*cotq2)) / (q*q))
+            c[2] = 0.5*x*beta*cg + 0.5*y*beta*sg + \
+                   z * (1 + ((1 - 0.5*q*cotq2) * (-beta*beta*cg*cg - \
+                        beta*beta*sg*sg)) / (q*q))
+            c[3] = beta * (-sg)
+            c[4] = beta * cg
+            c[5] = 0
+
+        with gil:
+
+            return np.array(c)
+
+    @cython.wraparound(False)
+    @cython.boundscheck(False)
+    @cython.nonecheck(False)
+    @cython.cdivision(True)
+    cdef double kernel(self, double [:] c) nogil:
+        """ Evaluate the kernel based on the coordinate map.
+        """
+        return 1 / (8*sqrt(2)) * \
+               sqrt(PI)*self.t*sqrt(self.t*self.D33)*sqrt(self.D33*self.D44) * \
+               1 / (16*PI*PI*self.D33*self.D33*self.D44*self.D44*self.t*self.t* \
+                    self.t*self.t) * \
+               exp(-sqrt((c[0]*c[0] + c[1]*c[1]) / (self.D33*self.D44) + \
+                   (c[2]*c[2] / self.D33 + (c[3]*c[3]+c[4]*c[4]) / self.D44) * \
+                   (c[2]*c[2] / self.D33 + (c[3]*c[3]+c[4]*c[4]) / self.D44) + \
+                    c[5]*c[5]/self.D44) / (4*self.t));
+
+cdef double PI = 3.1415926535897932
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+cdef double [:] euler_angles(double [:] input) nogil:
```

Comment:
`input` is a reserved name in Python. Please use something else for this variable name.